### PR TITLE
feat: add support for messageId

### LINKED ISSUE #414: Add support for messageId
#### Reason/Context
This is a follow-up issue coming from https://github.com/asyncapi/spec/issues/458.

#### Description
As per the RFC, we need to add support for `messageId` at message level. This issue helps tracking progress on it.
Note that it should not break any behavior in place, and the parser should be smart enough for detecting if you are using the new `messageId` field or rather relying on the name (no breaking change should be introduced, so it should be compatible with previous spec version)

Note: since the feature is not considered a breaking change, we could include it either on `2.3.0` or `3.0.0`, depending on how "fast" we move on.

cc @WaleedAshraf

### PR DESCRIPTION
**Pull request by WaleedAshraf**
**Description**
- add support for `messageId`
- If `messageId` is present, check for duplicates
- `messageId` takes precedence in `message.uid()` method
- update/add tests

**Related issue(s)**
closes #414

### CODE REVIEW

File: lib/customValidators.js [hunk 1]
```diff
@@ -174,6 +174,70 @@ function validateOperationId(
...
+function validateMessageId(
+  parsedJSON,
+  asyncapiYAMLorJSON,
+  initialFormat,
+  operations
+) {
+  const chnls = parsedJSON.channels;
+  if (!chnls) return true;
+  const chnlsMap = new Map(Object.entries(chnls));
+  //it is a map of paths, the one that is a duplicate and the one that is duplicated
```
Comment (magicmatatjahu) [inline lib/customValidators.js:195]: Fix that description 😄

File: lib/customValidators.js [hunk 2]
```diff
@@ -174,6 +174,70 @@ function validateOperationId(
...
+    const messageId = msg.messageId;
+    if (!messageId) return;
+
+    const messagePath = `${tilde(channelName)}/${opName}/message${oneOf}/messageId`;
+    const isMessageIdDuplicated = allMessages.find(v => v[0] === messageId);
+    if (!isMessageIdDuplicated)
+      return allMessages.push([messageId, messagePath]);
+
+    //isMessageIdDuplicated always holds one record and it is an array of paths, the one that is a duplicate and the one that is duplicated
+    duplicatedMessages.set(messagePath, isMessageIdDuplicated[1]);
```
Comment (magicmatatjahu) [inline lib/customValidators.js:210]: We should optimize it:

```suggestion
    const isMessageIdDuplicated = allMessages.find(v => v[0] === messageId);
    if (!isMessageIdDuplicated) {
      const messagePath = `${tilde(channelName)}/${opName}/message${oneOf}/messageId`;
      return allMessages.push([messageId, messagePath]);
    }

    //isMessageIdDuplicated always holds one record and it is an array of paths, the one that is a duplicate and the one that is duplicated
    duplicatedMessages.set(messagePath, isMessageIdDuplicated[1]);
```

Also `the one that is a duplicate and the one that is duplicated` 🤔  I'm not native in English but I think you've repeated yourself here.

File: lib/customValidators.js [hunk 3]
```diff
@@ -174,6 +174,70 @@ function validateOperationId(
...
+  };
+
+  chnlsMap.forEach((chnlObj, chnlName) => {
+    operations.forEach((opName) => {
+      const op = chnlObj[String(opName)];
+      if (op && op.message) {
+        if (op.message.oneOf) op.message.oneOf.forEach(msg => addDuplicateToMap(msg, chnlName, opName , '/oneOf'));
```
Comment (magicmatatjahu) [inline lib/customValidators.js:217]:
```suggestion
        if (op.message.oneOf) op.message.oneOf.forEach((msg, index)=> addDuplicateToMap(msg, chnlName, opName , `/oneOf/${index}`));
```

File: test/customValidators_test.js [hunk 4]
```diff
@@ -1076,6 +1077,273 @@ describe('validateOperationId()', function () {
...
+    const parsedInput = JSON.parse(inputString);
+
+    const expectedErrorObject = {
+      type: 'https://github.com/asyncapi/parser-js/validation-errors',
+      title: 'messageId must be unique across all the messages.',
+      parsedJSON: parsedInput,
+      validationErrors: [
+        {
+          title:
+            'test/2/subscribe/message/messageId is a duplicate of: test/1/publish/message/messageId',
```
Comment (magicmatatjahu) [inline test/customValidators_test.js:1157]: You should use tilde also in titles and full path

```suggestion
            '/channels/test~12/subscribe/message/messageId is a duplicate of: channels/test~11/publish/message/messageId',
```

etc. Please fix it in the `validateMessageId` function.

**Review by magicmatatjahu — CHANGES_REQUESTED** [review]
Overall good but we need improvements :)

File: lib/customValidators.js (hunk 2 above)
Comment (WaleedAshraf) [reply lib/customValidators.js:210]: Do you mean to move `messagePath` in the IF block? It is used on line 211.

File: lib/customValidators.js (hunk 3 above)
Comment (WaleedAshraf) [reply lib/customValidators.js:217]: What's the use of `index` here? 🤔

File: lib/customValidators.js [hunk 5]
```diff
@@ -174,6 +174,70 @@ function validateOperationId(
...
+/**
+ * Validates if messageIds are duplicated in the document
+ *
+ * @private
+ * @param  {Object} parsedJSON parsed AsyncAPI document
+ * @param  {String} asyncapiYAMLorJSON AsyncAPI document in string
+ * @param  {String} initialFormat information of the document was originally JSON or YAML
+ * @returns {Boolean} true in case the document is valid, otherwise throws {@link ParserError}
+ */
+function validateMessageId(
```
Comment (WaleedAshraf) [inline lib/customValidators.js:186]: @magicmatatjahu I see your suggestions on this method.
This method is a (almost) copy of the already existing method above in this file,

This method is copy of already existing https://github.com/asyncapi/parser-js/blob/d3dd11d885074c29ef5cde8956c5968351a54d7a/lib/customValidators.js#L123

I tried to keep it consistent with that. For example this comment :D
> it is a map of paths, the one that is a duplicate and the one that is duplicated

Let me know if we still need to make changes.

File: test/customValidators_test.js (hunk 4 above)
Comment (WaleedAshraf) [reply test/customValidators_test.js:1157]: We are using tilde for title, but it gets untidle in the `utils.groupValidationErrors` method.

https://github.com/asyncapi/parser-js/blob/d3dd11d885074c29ef5cde8956c5968351a54d7a/lib/utils.js#L259

File: lib/customValidators.js (hunk 2 above)
Comment (magicmatatjahu) [reply lib/customValidators.js:210]: Yeah... 🤦🏼 I wanted to be so cool and I came off as a dork 😆  You're right, your code is ok.

File: lib/customValidators.js (hunk 3 above)
Comment (magicmatatjahu) [reply lib/customValidators.js:217]: `message/oneOf` has a shape of array, so you need to clarify which element of array has potential problems :)

File: test/customValidators_test.js (hunk 4 above)
Comment (magicmatatjahu) [reply test/customValidators_test.js:1157]: Ok, I forgot about it, please go ahead and don't worry about my comment. We should fix it in the another PR.

File: lib/customValidators.js (hunk 5 above)
Comment (magicmatatjahu) [reply lib/customValidators.js:186]: Ok 😄  Please go ahead and don't worry about it. We should fix it in the another PR.

**Review by magicmatatjahu — APPROVED** [review]
LGTM!

**Comment by magicmatatjahu** [comment]
@smoya Could we merge it as next prerelease?

**Comment by smoya** [comment]
> @smoya Could we merge it as next prerelease?

There are pending things still to work on the spec PR https://github.com/asyncapi/spec/issues/458

**Comment by WaleedAshraf** [comment]
@smoya @magicmatatjahu It's showing changes in `.github` folder because `2022-04-release` branch is not up to date with `master`. Maybe you should do that? before merging this?

**Comment by WaleedAshraf** [comment]
link checker is failing because this link is updated,
https://apidevtools.org/json-schema-ref-parser/docs/options.html

to this,
https://apitools.dev/json-schema-ref-parser/docs/options.html

I'm not sure why it wasn't caught in recent [PRs](https://github.com/asyncapi/parser-js/pull/528) by @smoya 
Can someone fix this in the release branch? and I'll rebase.

**Comment by derberg** [comment]
Hey, since you edited API.md there are some broken links discovered with one of CI checks. There are 2 options:
- fix links
- or just remove any changes to API.md as we will anyway rebuild it by CI after release

**Comment by WaleedAshraf** [comment]
@derberg still fails. It seems the checker is using the base branch.

**Review by derberg — APPROVED** [review]
LGTM

@magicmatatjahu please also approve again, just to make sure you are ok as I see you revied and approved before

**Comment by magicmatatjahu** [comment]
@derberg Accepted. Please merge when you want.

**Comment by smoya** [comment]
Good news @WaleedAshraf! We are almost done with merging your feature into the release branch! 🎉 
However, for merging this PR, I need to ask you to:
- Fix the conflicts with the upstream branch (`2022-04-release`)
- Update `@asyncapi/specs` package to the latest prereleased version (which includes your changes). It is `v2.14.0-2022-04-release.3`. 

Once you push those changes, all will be ready to merge this. Thanks!

**Comment by WaleedAshraf** [comment]
@smoya Done 🎉

**Pull request merged**